src/lexparse/parser.go

```diff
@@ -0,0 +1,218 @@
+package lexparse
+
+import (
+  "fmt"
+  "strings"
+)
+
+type NodeType int
+
+const (
+  DeclarationS NodeType = iota
+  ExpressionS
+  TypeDefnS
+
+  ArrayTypeN
+  SetTypeN
+  MapTypeN
+  FunctionTypeN
+
+  DeclaredChainN
+  DeclaredN
+  DeclaredLeftN
+  FunctionArgdefN
+
+  SimpleExprN
+)
+
+var simpleExprTokenTypes map[tokenType]bool = map[tokenType]bool {
+  SymbolT: true,
+  StringT: true,
+  CharT: true,
+  IntT: true,
+  FloatT: true,
+  BoolT: true,
+  NullT: true,
+}
+
+// How the Node struct is used:
+//
+// token is the symbol token/int token/etc for SimpleExprN,
+// and the binop/comparison/setter/prefix/suffix token for those
+// nodetypes. Otherwise it's usually just the first token of
+// a node and not important aside for parse error printouts.
+//
+// left and right are obviously used in branching, but some nodetypes
+// only have one descendant, e.g. ExpressionS, PrefixN, SuffixN.
+// For node types with one descendant, only left is used.
+// SimpleExprN is the only node type that uses neither left or right.
+
+type Node struct {
+  ntype NodeType
+  token *Token
+  left *Node
+  right *Node
+}
+
+func (node Node) Printout(indent int) {
+  space := strings.Repeat(" ", indent*2)
+  fmt.Printf("%s%d\n", space, node.ntype)
+  fmt.Printf("%s%s\n", space, node.token.to_str())
+  if node.left != nil {
+    node.left.Printout(indent+1)
+  }
+  if node.right != nil {
+    node.right.Printout(indent+1)
+  }
+}
+
+type Parser struct {
+  tokens []Token
+  position int
+  typeLabels map[string]bool
```

Comment:
Why isn't this a global constant?

Reply:
What, `map[string]bool`?

Reply:
Just the `typeLabels` map overall.

Reply:
For the record, this was resolved in DMs. Part of the parser refactor will involve no longer needing type labels at all.